lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,418 @@
+        if (areAllUserReadFirstLane(inst))
+          continue;
+
+        applyReadFirstLane(inst, builder);
+        changed = true;
+      }
+    }
+  }
+  return changed;
+}
+
+// =====================================================================================================================
+// Collect the instructions that can be "assumed uniform" if the value itself isn't known to be uniform (by
+// DivergenceAnalysis), but we can replace all of its uses by readfirstlane(V) and still be correct
+//
+// @param block : The processing basic block
+// @param initialReadFirstLanes : The initial amdgcn_readfirstlane vector
+void PatchReadFirstLane::collectAssumeUniforms(BasicBlock *block,
+                                               const SmallVectorImpl<Instruction *> &initialReadFirstLanes) {
+  for (auto readfirstlane : initialReadFirstLanes)
+    m_canAssumeUniformDivergentUsesMap.insert({readfirstlane, {}});
+
+  for (Instruction &inst : llvm::reverse(*block)) {
+    // Skip readfirstlane-irrelevant instructions or a source of divergence which isn't clear to determine its operands
+    if (m_canAssumeUniformDivergentUsesMap.count(&inst) == 0 || (m_targetTransformInfo->isSourceOfDivergence(&inst)))
+      continue;
+
+    // For all operands, see if they can be assumed uniform
+    for (Use &use : inst.operands()) {
+      if (!m_divergenceAnalysis->isDivergentUse(&use))
+        continue;
+      auto operandInst = dyn_cast<Instruction>(use.get());
+      if (!operandInst)
+        continue;
+      if (m_canAssumeUniformDivergentUsesMap.count(operandInst) == 1)
+        continue;
+      if (operandInst->getParent() != block || !isAllUsersDefinedInBlock(operandInst, block))
+        continue;
+      // The operand can be assumed uniform becasue all users are assumed uniform
+      if (isAllUsersAssumedUniform(operandInst)) {
+        m_canAssumeUniformDivergentUsesMap[operandInst] = {};
+        m_canAssumeUniforms.push_back(operandInst);
+      }
+    }
+
+    bool canCollectOperands = true;
+    // NOTE: Stop propagation for a loadInst whose address space is fat pointer because it has uncertain size
+    auto loadInst = dyn_cast<LoadInst>(&inst);
+
+    // For all operands, see if all operands can be assumed uniform and are in the same block. If true, we can collect
+    // the operands to be used for the propagation heuristic in findInsertLocation.
+    SmallVector<Instruction *, 2> operandInsts;
+    for (Use &use : inst.operands()) {
+      if (auto operandInst = dyn_cast<Instruction>(use.get())) {
+        if (!m_canAssumeUniformDivergentUsesMap.count(operandInst))
+          continue;
+        if (operandInst->getParent() != block || !isAllUsersDefinedInBlock(operandInst, block) ||
+            (loadInst && loadInst->getPointerAddressSpace() == ADDR_SPACE_BUFFER_FAT_POINTER)) {
+          canCollectOperands = false;
+          break;
+        }
+        operandInsts.push_back(operandInst);
+      }
+    }
```

Comment:
IMHO, This loop can be merged with the above for-loop. I don't see any blocking reason. Maybe I missed something?

Reply:
Thanks. You're right. The two for-loops can be merged into one. Done.